### PR TITLE
Commcare ccz file downloads

### PR DESCRIPTION
**Pull request by proteusvacuum**
requires:
https://github.com/dimagi/django-soil/pull/22

@snopoke 
@gcapalbo

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/views.py, line 2473 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2503 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/views.py, line 272 - E501 line too long (122 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

File: corehq/apps/app_manager/templates/app_manager/download_index.html [hunk 1]
```diff
@@ -56,4 +56,7 @@
         });
     </script>
 </div>
+{% url "download_ccz" app.domain app.id as ccz_url %}
+{% include 'app_manager/partials/download_async_modal.html' with element_id=download_ccz_id download_url=ccz_url%}
```
Comment (snopoke) [inline corehq/apps/app_manager/templates/app_manager/download_index.html]: nice

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/views.py, line 2473 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2503 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/views.py, line 272 - E501 line too long (122 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

File: corehq/apps/app_manager/views.py [hunk 2]
```diff
@@ -2474,39 +2466,65 @@ def download_index(request, domain, app_id, template="app_manager/download_index
...
+        )
+
+    return _files(files), errors
+
+
 class DownloadCCZ(DownloadMultimediaZip):
     name = 'download_ccz'
     compress_zip = True
     zip_name = 'commcare.ccz'
+    download_async = True
```
Comment (snopoke) [inline corehq/apps/app_manager/views.py]: this doesn't seem to be used anywhere

File: corehq/util/zip_utils.py [hunk 3]
```diff
@@ -66,3 +73,74 @@ def get(self, request, *args, **kwargs):
...
+                                download_id=download_id,
+                                content_disposition='attachment; filename="{fname}"'.format(fname=filename),
+                                use_transfer=use_transfer), errors
+
+
+class DownloadZipAsync(DownloadZip):
+    include_multimedia_files = False
+    include_index_files = False
+
+    def get(self, request, *args, **kwargs):
```
Comment (snopoke) [inline corehq/util/zip_utils.py]: Should add an assert here that either `include_multimedia_files` or `include_index_files` is true

Comment (snopoke) [inline corehq/util/zip_utils.py]: or both

File: corehq/util/zip_utils.py [hunk 4]
```diff
@@ -66,3 +73,74 @@ def get(self, request, *args, **kwargs):
...
+        file_iterator, errors = iter_media_files(app.get_media_objects())
+    if include_index_files:
+        from corehq.apps.app_manager.views import iter_index_files
+        if app.is_remote_app():
+            file_iterator, errors = iter_index_files(app)
+        else:
+            index_files, index_file_errors = iter_index_files(app)
+            if index_file_errors:
+                errors.append(index_file_errors)
+            file_iterator = itertools.chain(file_iterator, index_files)
```
Comment (snopoke) [inline corehq/util/zip_utils.py]: I think the `is_remote` check here is in the wrong place. The logic should be: only include multimedia if app is not remote

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/views.py, line 2473 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2503 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/views.py, line 279 - E501 line too long (122 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

**Comment by gcapalbo** [comment]
build failure looks related to removing zip_utils

**Comment by dannyroberts** [comment]
@proteusvacuum can you link to a ticket or add a description to the PR? Would be curious to follow along

**Comment by proteusvacuum** [comment]
@dannyroberts, it's related to this: http://manage.dimagi.com/default.asp?167711

This PR moves ccz and multimedia zip file generation into an async celery task.

- Additionally, it checks to see if files are of a multimedia type and doesn't try to compress these (as this takes ++ time, and doesn't lead to very much reduction in file size. This was already happening for multimedia only zips, but for ccz files, they were being compressed.)
- It also uses @snopoke's django_transfer addition to store the file directly on the NFS drive. This leads to a few advantages:
  - we aren't putting 500+MB binary blobs into redis
  - if this file already exists on the nfs drive, it doesn't recreate it, and serves it immediately

**Comment by dimagimon** [comment]
There are 7 errors:

* corehq/apps/app_manager/views.py, line 2478 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2508 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/views.py, line 12 - F811 redefinition of unused 'settings' from line 7
* corehq/apps/hqmedia/views.py, line 27 - F401 'transfer_enabled' imported but unused
* corehq/apps/hqmedia/views.py, line 293 - E501 line too long (126 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/app_manager/views.py, line 2478 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2508 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/views.py, line 12 - F811 redefinition of unused 'settings' from line 7
* corehq/apps/hqmedia/views.py, line 291 - E501 line too long (126 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/app_manager/views.py, line 2478 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2508 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/views.py, line 13 - F811 redefinition of unused 'settings' from line 8
* corehq/apps/hqmedia/views.py, line 292 - E501 line too long (126 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/app_manager/views.py, line 2478 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2508 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/tasks.py, line 110 - E127 continuation line over-indented for visual indent
* corehq/apps/hqmedia/views.py, line 288 - E501 line too long (126 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

**Comment by snopoke** [comment]
@proteusvacuum I fixed a few small things and made one or two changes (just moving stuff around). Wasn't sure when you'll get back to this and quite keen to get it merged in.

**Comment by dimagimon** [comment]
There are 7 errors:

* corehq/apps/app_manager/views.py, line 2478 - E126 continuation line over-indented for hanging indent
* corehq/apps/app_manager/views.py, line 2508 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqmedia/tasks.py, line 109 - E127 continuation line over-indented for visual indent
* corehq/apps/hqmedia/tasks.py, line 111 - W293 blank line contains whitespace
* corehq/apps/hqmedia/views.py, line 288 - E501 line too long (126 > 115 characters)
* corehq/apps/importer/views.py, line 68 - E226 missing whitespace around arithmetic operator
* corehq/apps/users/views/mobile/users.py, line 763 - E226 missing whitespace around arithmetic operator

**Comment by proteusvacuum** [comment]
LGTM, thanks!

**Comment by gcapalbo** [comment]
@snopoke @proteusvacuum sounds like these three PRs are ready to go - do they need to be qa'd at all, or are they good to merge?

https://github.com/dimagi/django-soil/pull/22
https://github.com/dimagi/couchexport/pull/135
https://github.com/dimagi/commcare-hq/pull/6712

**Comment by snopoke** [comment]
@gcapalbo I think they are gtg

**Pull request merged**